### PR TITLE
[Graph][Model] Cache adj & inc; MX GCN

### PR DESCRIPTION
**Pull request by jermainewang**
## Description

Cache adj and inc matrix in graph index so the construction overhead can be saved for static graph. Expose a new interface `clear_cache` for users if memory is a concern. For GCN model, try to match Kipf's implementation in both pytorch and MXNet. Managed to produce the same test accuracy ~81%.  For speed, on my K20, torch's version is ~4x slower than MX's. I guess this is because we didn't use the coalesced spmm. Will fix this later.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

## Changes

- [x] Cache adj & inc  in graph index.
- [x] Fix `pytorch/gcn/gcn_spmv.py`
- [x] Add `mxnet/gcn/gcn_spmv.py`
~~Fix `mxnet/gcn/gcn_batch.py`.~~ will be done by @ZiyueHuang
~~Fix `pytorch/gcn/gcn_batch.py`.~~ will be done by @ZiyueHuang

### CODE REVIEW

**Comment by jermainewang** [comment]
@ZiyueHuang

File: examples/mxnet/gcn/gcn_spmv.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                 g,
+                 in_feats,
+                 out_feats,
+                 activation,
+                 dropout,
+                 bias=True):
+        super(GCNLayer, self).__init__()
+        self.g = g
+        with self.name_scope():
+            self.dense = gluon.nn.Dense(out_feats, use_bias=bias,
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply examples/mxnet/gcn/gcn_spmv.py]: Yeah, I'm aware of that. I don't know how to register raw parameter tensors in gluon (things equivalent to `torch.nn.Parameter`). @yifeim do you know?

[the remark this answers is not in the document]
Comment (szha) [reply examples/mxnet/gcn/gcn_spmv.py]: For Block, you need to manually call `self.parameter_name.data(ctx)`. This is only automated for HybridBlock.

[the remark this answers is not in the document]
Comment (jermainewang) [reply examples/mxnet/gcn/gcn_spmv.py]: Ok, I looked at the source code in `gluon.Dense` and used `self.params.get`. Don't know whether that's the best practice.

File: examples/mxnet/gcn/gcn_spmv.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        super(GCNLayer, self).__init__()
+        self.g = g
+        with self.name_scope():
+            self.weight = self.params.get('weight', shape=(in_feats, out_feats),
+                    init=mx.init.Xavier())
+            if bias:
+                self.bias = self.params.get('bias', shape=(out_feats,),
+                    init=mx.init.Zero())
+            else:
+                self.bias = None
```
Comment (zheng-da) [inline examples/mxnet/gcn/gcn_spmv.py]: why not use gluon Linear layer?

File: python/dgl/graph.py [hunk 3]
```diff
@@ -385,6 +385,15 @@ def clear(self):
         self._msg_graph.clear()
         self._msg_frame.clear()
 
+    def clear_cache(self):
```
Comment (zheng-da) [inline python/dgl/graph.py]: I think it's better to make the name very specific for clearing cache for graph structures. Maybe passing an argument?

File: examples/mxnet/gcn/gcn_spmv.py (hunk 2 above)
Comment (jermainewang) [reply examples/mxnet/gcn/gcn_spmv.py]: Simply follows the codes here: https://github.com/tkipf/pygcn/blob/4396e4db5b97c9e071516bc601c9b6693696c489/pygcn/layers.py#L31-L37 . The weight is used before message passing while the bias is used afterwards.

File: python/dgl/graph.py (hunk 3 above)
Comment (jermainewang) [reply python/dgl/graph.py]: How about we adding extra option arguments when we have more types of caches in the future? Currently, I'm not sure what are the other cache types, and I want to keep it the API as simple as possible at the moment.

File: examples/mxnet/gcn/gcn_spmv.py (hunk 2 above)
Comment (zheng-da) [reply examples/mxnet/gcn/gcn_spmv.py]: interesting. do you know any reason why it should be done like this?

File: python/dgl/graph.py (hunk 3 above)
Comment (zheng-da) [reply python/dgl/graph.py]: i think the main concern is back compatibility. Adding an argument in the future is kind of breaking API.

I'm sure we'll have other cache types. What I'm working on is the GPU caching for node embeddings. We'll need to provide caching in the main memory in the distributed setting.

File: examples/mxnet/gcn/gcn_spmv.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                 n_hidden,
+                 n_classes,
+                 n_layers,
+                 activation,
+                 dropout,
+                 normalization):
+        super(GCN, self).__init__()
+        self.layers = gluon.nn.Sequential()
+        # input layer
+        self.layers.add(GCNLayer(g, in_feats, n_hidden, activation, dropout))
```
Comment (ZiyueHuang) [inline examples/mxnet/gcn/gcn_spmv.py]: Should `dropout=0` when applied on the input features? According to https://github.com/tkipf/pygcn/blob/master/pygcn/models.py.

File: python/dgl/graph.py (hunk 3 above)
Comment (jermainewang) [reply python/dgl/graph.py]: Breaking API at the current stage is fine. This is definitely not the final API in our next release. For distributed setting, we will likely to have extra APIs (similar to DataParallelTrainer). We could put your caching option there. I would suggest modify this API again (if necessary) in the PR for distributed training, so we can have the whole picture to look at.

File: examples/mxnet/gcn/gcn_spmv.py (hunk 2 above)
Comment (jermainewang) [reply examples/mxnet/gcn/gcn_spmv.py]: The math is `relu(A * H * W + b)`, where `A` is the normalized adj, `H` is the feature, `W` is the weight, `b` is the bias. There could be two implementations: (1) do message passing first (A * H), then apply a dense layer. (2) Do H * W first, then message passing, then add bias. I guess why the author chose the later one is because the input dimension is much bigger than the hidden dimension, so H*W first makes the computation more efficient.

File: examples/mxnet/gcn/gcn_spmv.py (hunk 4 above)
Comment (jermainewang) [reply examples/mxnet/gcn/gcn_spmv.py]: I see. Thx, will fix that.

**Comment by yifeim** [comment]
"Managed to produce the same test accuracy ~81%." Since there seem to be test coverages already, I would be happy if the accuracy claim could be reflected in the documents.

File: examples/mxnet/gcn/gcn_spmv.py (hunk 2 above)
Comment (zheng-da) [reply examples/mxnet/gcn/gcn_spmv.py]: i see. thanks for the explanation.

File: python/dgl/graph.py (hunk 3 above)
Comment (zheng-da) [reply python/dgl/graph.py]: sounds good.
